<a href="https://colab.research.google.com/github/Kavia-M/Data_science_course-Intellipaat/blob/main/Python/Assignments/Assignment_Quiz/Random_Forest_Assignment_Quiz.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**Problem Statement:**

1. Census-income data plays the most important role in the democratic system of government, highly affecting the economic sectors. Census-related figures are used to allocate federal funding by the government to different states and localities.

2. Census data is also used for post census residents estimates and predictions, economic and social science research, and many other such applications. Therefore, the importance of this data and its accurate predictions is very clear to us. The main aim is to increase awareness about how the income factor actually has an impact not only on the individual lives of citizens but also an effect on the nation and its betterment. You will have a look at the data pulled out from the 1994 Census bureau database, and try to find insights into how various features have an effect on the income of an individual.

3. The data contains approximately 32,000 observations with over 15 variables.

4. The strategy is to analyze the data and perform a predictive task of classification to predict whether an individual makes over 50K a year or less by using a logistic regression algorithm. .

**Data Description:**


| Column Names      | Description                                          |
|-------------------|-----------------------------------------------------|
| Age               | Age of the individual                               |
| Workclass         | department of the working individual                |
| fnlwgt            | Final weight of the individual                      |
| education         | The education degree of the individual              |
| education-num     | Number of years of education                        |
| marital-status    | Marital status of the individual                    |
| occupation        | Occupation of the individual                        |
| relationship      | Relation value                                      |
| race              | Ethnicity of the individual                         |
| sex               | Female, Male                                        |
| capital-gain      | capital gain of the individual                      |
| capital-loss      | capital loss of the individual                      |
| hours-per-week    | number of working hours                             |
| native-country    | The native country of the individual                |
| Annual-Income     | Annual income either >50K or <=50K                  |

Both the datasets iven in the download zip file are already used in Logistic Regression Quizzes

1. [Logistic_Regression_Assignment_Quiz.ipynb](https://github.com/Kavia-M/Data_science_course-Intellipaat/blob/main/Python/Assignments/Assignment_Quiz/Logistic_Regression_Assignment_Quiz.ipynb) - census-income.csv

2. [Quiz_Logistic_regression_(live_class).ipynb](https://github.com/Kavia-M/Data_science_course-Intellipaat/blob/main/Python/Assignments/Assignment_Quiz/Quiz_Logistic_regression_(live_class).ipynb) - Airport_Passengers_Satisfaction.csv as Invistico_Airline.csv

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
census = pd.read_csv('/content/census-income.csv')

1) What is the biggest advantage that helps random forest classifiers to triumph over the decision trees?
- It Combines all positive predictions from all decision trees
- All of the above
- It works on the bagging method(bootstrap method)
- It has shown great predictive results over decision tree models.

It takes majority vote only but not exactly combining all positive results

It is Bagging

It has shown better predictive results than decision trees

1. Answer - All of the above

2) In a given problem where you have a very large dataset with both continuous and categorical features, why would you choose the random forest classifier?
- Works well with the high dimensional data
- High accuracy with less need for interpretation
- All of the above
- Random forest can work on both regression and classification problem

For this question, the correct choice is:

**✅ All of the above**

Because:

* **Works well with high dimensional data** → Yes, Random Forest handles many features (both continuous & categorical) without much preprocessing.
* **High accuracy with less need for interpretation** → Yes, Random Forest generally gives high accuracy but is not very interpretable compared to a single decision tree.  - Yes. single dtree is more understandable by us
* **Can work on both regression and classification problems** → Yes, that’s a big advantage — it’s versatile. since it used dtrees, and dtrees can work with both kind of data

So the best answer is **All of the above**.



2. Answer - All of the above

3) Which of the following techniques is used in the Random Forest model?
- Boosting
- None of the these
- Bagging
- Ensemble


* **Ensemble** = a general term for combining multiple models to improve performance.

  * Two main types:

    * **Bagging (Bootstrap Aggregation)** → Random Forest uses this.
    * **Boosting** → e.g., AdaBoost, Gradient Boosting, XGBoost.

* **Random Forest specifically uses Bagging** ✅

  * Each tree is trained on a bootstrap sample (random subset of rows, with replacement).
  * At each split, only a random subset of features is considered.

Since Bagging is more accurate than ensemble, we choose that answer

3. Answer - Bagging

4)
Choose the percentage of population with income greater than 50K income?
- 75.5%
- 25.2%
- 24.8%
- 35.5%

In [3]:
census.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  Unnamed: 14     32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [4]:
census

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,Unnamed: 14
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32557,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32558,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K
32559,22,Private,201490,HS-grad,9,Never-married,Adm-clerical,Own-child,White,Male,0,0,20,United-States,<=50K


Last column header missing in csv

In [5]:
census.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'Unnamed: 14'],
      dtype='object')

In [6]:
census.columns[:-1] # end exclusive, stop before last element

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country'],
      dtype='object')

In [7]:
type(census.columns[:-1])

pandas.core.indexes.base.Index

In [8]:
type(census.columns[:-1].to_numpy())

numpy.ndarray

In [9]:
np.append(census.columns[:-1].to_numpy(), 'annual_income')

array(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'annual_income'], dtype=object)

In [10]:
census.columns = np.append(census.columns[:-1].to_numpy(), 'annual_income')

In [11]:
census.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'annual_income'],
      dtype='object')

In [12]:
census.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  annual_income   32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [13]:
census

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,annual_income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32557,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32558,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K
32559,22,Private,201490,HS-grad,9,Never-married,Adm-clerical,Own-child,White,Male,0,0,20,United-States,<=50K


In [14]:
census.isnull().sum().sum()

np.int64(0)

In [15]:
census.duplicated().sum()

np.int64(24)

In [16]:
census_without_duplicates = census.drop_duplicates()

In [17]:
census.duplicated().sum()

np.int64(24)

In [18]:
census_without_duplicates.duplicated().sum()

np.int64(0)

In [19]:
census['annual_income'].value_counts()

,count
annual_income,
<=50K,24720
>50K,7841


In [20]:
# to get probability

census['annual_income'].value_counts(normalize=True)

,proportion
annual_income,
<=50K,0.75919
>50K,0.24081


In [21]:
census['annual_income'].value_counts(normalize=True)*100

,proportion
annual_income,
<=50K,75.919044
>50K,24.080956


In [22]:
census['annual_income'].value_counts(normalize=True)['>50K']

np.float64(0.2408095574460244)

In [23]:
type(census['annual_income'].value_counts(normalize=True))

pandas.core.series.Series

In [24]:
type(census['annual_income'].value_counts(normalize=True)*100)

pandas.core.series.Series

In [25]:
(census['annual_income'].value_counts(normalize=True)*100)['>50K']

np.float64(24.080955744602438)

24.08% is accurate but 24.8% is closer in given answers

In [26]:
(census_without_duplicates['annual_income'].value_counts(normalize=True)*100)['>50K']

np.float64(24.092571533945968)

4. Answer - 24.8%

5) Compute how many samples of the population are never married and working hours less than 20 hours?
- 123
- 127
- 785
- 134

In [27]:
census['marital-status'].unique()

array(['Never-married', 'Married-civ-spouse', 'Divorced',
       'Married-spouse-absent', 'Separated', 'Married-AF-spouse',
       'Widowed'], dtype=object)

In [28]:
census['workclass'].unique()

array(['State-gov', 'Self-emp-not-inc', 'Private', 'Federal-gov',
       'Local-gov', '?', 'Self-emp-inc', 'Without-pay', 'Never-worked'],
      dtype=object)

In [29]:
census['hours-per-week'].dtype

dtype('int64')

In [30]:
census[(census['marital-status']=='Never-married') & (census['hours-per-week']<20)]

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,annual_income
168,18,Private,25828,11th,7,Never-married,Handlers-cleaners,Own-child,White,Male,0,0,16,United-States,<=50K
178,18,Private,183930,HS-grad,9,Never-married,Other-service,Own-child,White,Male,0,0,12,United-States,<=50K
209,17,Private,65368,11th,7,Never-married,Sales,Own-child,White,Female,0,0,12,United-States,<=50K
262,17,Private,245918,11th,7,Never-married,Other-service,Own-child,White,Male,0,0,12,United-States,<=50K
280,22,Private,34918,Bachelors,13,Never-married,Prof-specialty,Not-in-family,White,Female,0,0,15,Germany,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32345,18,Private,347336,12th,8,Never-married,Other-service,Own-child,White,Male,0,0,12,United-States,<=50K
32356,23,Private,133355,Some-college,10,Never-married,Adm-clerical,Own-child,White,Male,0,0,15,United-States,<=50K
32494,82,?,403910,HS-grad,9,Never-married,?,Not-in-family,White,Male,0,0,3,United-States,<=50K
32524,26,Private,191648,Assoc-acdm,12,Never-married,Machine-op-inspct,Other-relative,White,Female,0,0,15,United-States,<=50K


In [31]:
len(census[(census['marital-status']=='Never-married') & (census['hours-per-week']<20)])

907

In [32]:
(census=='?').sum()

,0
age,0
workclass,1836
fnlwgt,0
education,0
education-num,0
marital-status,0
occupation,1843
relationship,0
race,0
sex,0


In [33]:
census_without_missing = census.replace("?", np.nan)

In [34]:
(census=='?').sum()

,0
age,0
workclass,1836
fnlwgt,0
education,0
education-num,0
marital-status,0
occupation,1843
relationship,0
race,0
sex,0


In [35]:
census_without_missing.isnull().sum()

,0
age,0
workclass,1836
fnlwgt,0
education,0
education-num,0
marital-status,0
occupation,1843
relationship,0
race,0
sex,0


In [36]:
census_without_missing.dropna(inplace=True)

In [37]:
census_without_missing.isnull().sum()

,0
age,0
workclass,0
fnlwgt,0
education,0
education-num,0
marital-status,0
occupation,0
relationship,0
race,0
sex,0


In [38]:
census_without_missing.duplicated().sum()

np.int64(23)

In [39]:
census_without_dup_without_missing = census_without_missing.drop_duplicates()

In [40]:
len(census_without_missing[(census_without_missing['marital-status']=='Never-married')
& (census_without_missing['hours-per-week']<20)])

785

In [41]:
len(census_without_dup_without_missing[(census_without_dup_without_missing['marital-status']=='Never-married')
& (census_without_dup_without_missing['hours-per-week']<20)])

781

In [42]:
len(census_without_missing)

30162

In [43]:
len(census)

32561

**census_without_missing is mostly used**

In [44]:
len(census_without_missing[(census_without_missing['marital-status']=='Never-married')
& (census_without_missing['hours-per-week']<20)])

785

5. Answer - 785

6) Choose the correct list of age with minimum age , maximum and 50th Percentile of the age group?
- All
- [15,95,37]
- [17,90,36]
- [17,90,37]

They ask for minimum age, maximum age, 50th percentile of age group

In [45]:
census_without_missing['age'].min()

17

In [46]:
census_without_missing['age'].max()

90

In [47]:
census_without_missing['age'].quantile(0.5)  #quantile means use 0 to 1

np.float64(37.0)

In [48]:
# 50th percentile means median

census_without_missing['age'].median()

37.0

In [49]:
census['age'].median()

37.0

The **percentile** means "the value below which a certain percentage of the data falls."

* **50th percentile** → the value below which **50% of the data** lies.
* That’s exactly the definition of the **median**.

---

👉 Example:
Data: `[10, 20, 30, 40, 50]`

* **Median** = the middle value when sorted = `30`.
* **50th percentile** = half the values (50%) are below `30`, half are above → also `30`.

So:

* 25th percentile = first quartile (Q1).
* 50th percentile = median (Q2).
* 75th percentile = third quartile (Q3).

---


In [50]:
np.percentile(census['age'], 50)

np.float64(37.0)

In [51]:
np.quantile(census['age'], 0.5)

np.float64(37.0)

In [52]:
[census_without_missing['age'].min(),
 census_without_missing['age'].max(),
 int(census_without_missing['age'].quantile(0.5))]

[17, 90, 37]

In [53]:
[census['age'].min(),
 census['age'].max(),
 int(census['age'].quantile(0.5))]

[17, 90, 37]

6. Answer - [17,90,37]

7) From above census data which country has the highest population and  the lowest population?
- United-States and Holland-Netherlands
- United-States and scotland
- Mexico and Holland-Netherlands
- Scotland and Holland-Netherlands

In [54]:
census_without_missing.columns

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'annual_income'],
      dtype='object')

Population is not a column. in census data, number of rows is population

In [55]:
census['native-country'].value_counts()

,count
native-country,
United-States,29170
Mexico,643
?,583
Philippines,198
Germany,137
Canada,121
Puerto-Rico,114
El-Salvador,106
India,100


In [56]:
census['native-country'].value_counts().idxmin()

'Holand-Netherlands'

In [57]:
census['native-country'].value_counts().idxmax()

'United-States'

In [58]:
census_without_missing['native-country'].value_counts()

,count
native-country,
United-States,27504
Mexico,610
Philippines,188
Germany,128
Puerto-Rico,109
Canada,107
El-Salvador,100
India,100
Cuba,92


In [59]:
census_without_missing['native-country'].value_counts().idxmin()

'Holand-Netherlands'

In [60]:
census_without_missing['native-country'].value_counts().idxmax()

'United-States'

United-States is max population and Holand-Netherlands is min population in this data

7. Answer - United-States and Holland-Netherlands

8) How does n_estimators work in the random forest classifier?
- Number of decision trees
- Training epochs
- Number of iterations
- Number of random forests for the classifier.

n_estimators is number of decision trees

Parameters:
n_estimatorsint, default=100
The number of trees in the forest.

[scikit learn RandomForestClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html)

8. Answer - Number of decision trees

9) Can the target data for the random forest model be categorical or continuous value?
- No
- Yes

Yes. Since Random forest uses dtrees and Dtree can work on both categorical and continuous variables as target. Random forest also has the capability

9. Answer - Yes

10) How can you use hyperparameter tuning to your advantage while working with the random forest classifier?
- All of the above
- Improve the model’s performance
- Standardization of the data
- Normalizes the features

Improves model performance is the use of hyperparameter tuning

Random forest's parameters are (as per link above)

`class sklearn.ensemble.RandomForestClassifier(n_estimators=100, *, criterion='gini', max_depth=None, min_samples_split=2, min_samples_leaf=1, min_weight_fraction_leaf=0.0, max_features='sqrt', max_leaf_nodes=None, min_impurity_decrease=0.0, bootstrap=True, oob_score=False, n_jobs=None, random_state=None, verbose=0, warm_start=False, class_weight=None, ccp_alpha=0.0, max_samples=None, monotonic_cst=None)`

Normalization and Standardization are done in preprocessing and not in hyperparameter tuning

Also,

❌ Standardization or normalization is not required for Random Forest, because:

Decision trees (and forests) are scale-invariant.

They don’t care about the range of features.

So the main advantage of hyperparameter tuning is to improve model performance, not to normalize or standardize data.

10. Answer - Improve the model’s performance

11)
Select the best hyperparameters by RandomSearchCV and fit the model with the best hyperparameters and compute the accuracy score of the model.
- None of the above
- 85% and above
- 50% to 70%
- 30% to 50%

Seems like RandomForestClassifier takes only numeric values as feature values

In [77]:
census_without_missing.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30162 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             30162 non-null  int64 
 1   workclass       30162 non-null  object
 2   fnlwgt          30162 non-null  int64 
 3   education       30162 non-null  object
 4   education-num   30162 non-null  int64 
 5   marital-status  30162 non-null  object
 6   occupation      30162 non-null  object
 7   relationship    30162 non-null  object
 8   race            30162 non-null  object
 9   sex             30162 non-null  object
 10  capital-gain    30162 non-null  int64 
 11  capital-loss    30162 non-null  int64 
 12  hours-per-week  30162 non-null  int64 
 13  native-country  30162 non-null  object
 14  annual_income   30162 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [78]:
from sklearn.preprocessing import LabelEncoder

In [90]:
col_list_census = []

for i in census_without_missing.columns:
  if ((census_without_missing[i].dtypes =='object')& (i != 'annual_income')):
    col_list_census.append(i)

In [91]:
col_list_census

['workclass',
 'education',
 'marital-status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'native-country']

In [88]:
census_without_missing_encoded = census_without_missing.copy()
encoders_census = {}
mappings_census = {}
for i in census_without_missing_encoded.columns:
  if census_without_missing_encoded[i].dtype == 'object' and i!='annual_income':
    le = LabelEncoder() # new instance every time
    census_without_missing_encoded[i] = le.fit_transform(census_without_missing_encoded[i])
    encoders_census[i] = le # storing the correct encoder for each column
    mappings_census[i] = dict(zip(le.classes_,le.transform(le.classes_)))

In [92]:
mappings_census

{'workclass': {'Federal-gov': np.int64(0),
  'Local-gov': np.int64(1),
  'Private': np.int64(2),
  'Self-emp-inc': np.int64(3),
  'Self-emp-not-inc': np.int64(4),
  'State-gov': np.int64(5),
  'Without-pay': np.int64(6)},
 'education': {'10th': np.int64(0),
  '11th': np.int64(1),
  '12th': np.int64(2),
  '1st-4th': np.int64(3),
  '5th-6th': np.int64(4),
  '7th-8th': np.int64(5),
  '9th': np.int64(6),
  'Assoc-acdm': np.int64(7),
  'Assoc-voc': np.int64(8),
  'Bachelors': np.int64(9),
  'Doctorate': np.int64(10),
  'HS-grad': np.int64(11),
  'Masters': np.int64(12),
  'Preschool': np.int64(13),
  'Prof-school': np.int64(14),
  'Some-college': np.int64(15)},
 'marital-status': {'Divorced': np.int64(0),
  'Married-AF-spouse': np.int64(1),
  'Married-civ-spouse': np.int64(2),
  'Married-spouse-absent': np.int64(3),
  'Never-married': np.int64(4),
  'Separated': np.int64(5),
  'Widowed': np.int64(6)},
 'occupation': {'Adm-clerical': np.int64(0),
  'Armed-Forces': np.int64(1),
  'Craft-repai

In [93]:
census_without_missing

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,annual_income
0,39,State-gov,77516,Bachelors,13,Never-married,Adm-clerical,Not-in-family,White,Male,2174,0,40,United-States,<=50K
1,50,Self-emp-not-inc,83311,Bachelors,13,Married-civ-spouse,Exec-managerial,Husband,White,Male,0,0,13,United-States,<=50K
2,38,Private,215646,HS-grad,9,Divorced,Handlers-cleaners,Not-in-family,White,Male,0,0,40,United-States,<=50K
3,53,Private,234721,11th,7,Married-civ-spouse,Handlers-cleaners,Husband,Black,Male,0,0,40,United-States,<=50K
4,28,Private,338409,Bachelors,13,Married-civ-spouse,Prof-specialty,Wife,Black,Female,0,0,40,Cuba,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,27,Private,257302,Assoc-acdm,12,Married-civ-spouse,Tech-support,Wife,White,Female,0,0,38,United-States,<=50K
32557,40,Private,154374,HS-grad,9,Married-civ-spouse,Machine-op-inspct,Husband,White,Male,0,0,40,United-States,>50K
32558,58,Private,151910,HS-grad,9,Widowed,Adm-clerical,Unmarried,White,Female,0,0,40,United-States,<=50K
32559,22,Private,201490,HS-grad,9,Never-married,Adm-clerical,Own-child,White,Male,0,0,20,United-States,<=50K


In [94]:
census_without_missing_encoded

,age,workclass,fnlwgt,education,education-num,marital-status,occupation,relationship,race,sex,capital-gain,capital-loss,hours-per-week,native-country,annual_income
0,39,5,77516,9,13,4,0,1,4,1,2174,0,40,38,<=50K
1,50,4,83311,9,13,2,3,0,4,1,0,0,13,38,<=50K
2,38,2,215646,11,9,0,5,1,4,1,0,0,40,38,<=50K
3,53,2,234721,1,7,2,5,0,2,1,0,0,40,38,<=50K
4,28,2,338409,9,13,2,9,5,2,0,0,0,40,4,<=50K
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
32556,27,2,257302,7,12,2,12,5,4,0,0,0,38,38,<=50K
32557,40,2,154374,11,9,2,6,0,4,1,0,0,40,38,>50K
32558,58,2,151910,11,9,6,0,4,4,0,0,0,40,38,<=50K
32559,22,2,201490,11,9,4,0,3,4,1,0,0,20,38,<=50K


In [95]:
mappings_census.keys()

dict_keys(['workclass', 'education', 'marital-status', 'occupation', 'relationship', 'race', 'sex', 'native-country'])

In [96]:
list(mappings_census.keys()) == col_list_census

True

In [87]:
from sklearn.model_selection import train_test_split

In [97]:
X_census = census_without_missing_encoded.drop(['annual_income'], axis=1)

In [98]:
census_without_missing_encoded.shape

(30162, 15)

In [99]:
X_census.shape

(30162, 14)

In [100]:
y_census = census_without_missing_encoded['annual_income']

In [101]:
y_census.shape

(30162,)

In [102]:
X_train_census, X_test_census, y_train_census, y_test_census = train_test_split(
    X_census, y_census, test_size=0.3, random_state=42
)

In [103]:
X_train_census.shape

(21113, 14)

In [104]:
X_test_census.shape

(9049, 14)

In [70]:
y_train_census.shape

(21113,)

In [105]:
y_test_census.shape

(9049,)

In [106]:
from sklearn.ensemble import RandomForestClassifier

In [107]:
from sklearn.model_selection import RandomizedSearchCV

In [113]:
from sklearn.metrics import accuracy_score, confusion_matrix, precision_score, recall_score

In [108]:
rf = RandomForestClassifier()

# Hyperparameter grid
param_dist = {
    'n_estimators': [100, 200, 300, 400, 500],
    'max_depth': [None, 5, 10, 20, 30, 40, 50],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'max_features': [None, 'sqrt', 'log2'],
    'random_state': [0, 1, 5, 6, 24, 42],
    'criterion': ['gini', 'entropy'],
    'bootstrap': [True, False]
}

In [110]:
rf.fit(X_train_census, y_train_census)

RandomForestClassifier()

In [111]:
y_pred_census = rf.predict(X_test_census)

In [114]:
accuracy_score(y_test_census, y_pred_census)

0.8557851696320036

In [109]:
random_search = RandomizedSearchCV(
    estimator=rf,
    param_distributions=param_dist,
    n_iter=20,          # number of random combinations to try
    cv=5,               # 5-fold cross-validation
    scoring='accuracy',
    random_state=42,
    n_jobs=-1
)

In [116]:
random_search.fit(X_train_census, y_train_census)

RandomizedSearchCV(cv=5, estimator=RandomForestClassifier(), n_iter=20,
                   n_jobs=-1,
                   param_distributions={'bootstrap': [True, False],
                                        'criterion': ['gini', 'entropy'],
                                        'max_depth': [None, 5, 10, 20, 30, 40,
                                                      50],
                                        'max_features': [None, 'sqrt', 'log2'],
                                        'min_samples_leaf': [1, 2, 4],
                                        'min_samples_split': [2, 5, 10],
                                        'n_estimators': [100, 200, 300, 400,
                                                         500],
                                        'random_state': [0, 1, 5, 6, 24, 42]},
                   random_state=42, scoring='accuracy')

In [117]:
random_search.best_params_

{'random_state': 5,
 'n_estimators': 400,
 'min_samples_split': 10,
 'min_samples_leaf': 4,
 'max_features': 'sqrt',
 'max_depth': 50,
 'criterion': 'gini',
 'bootstrap': True}

In [118]:
# we can directly predict with the best params

y_pred_census1 = random_search.predict(X_test_census)

In [120]:
accuracy_score(y_test_census, y_pred_census1)

0.862084208199801

only a bit improved

In [121]:
# by applying the best params mannually

rf1 = RandomForestClassifier(
    random_state = 5,
 n_estimators = 400,
 min_samples_split = 10,
 min_samples_leaf = 4,
 max_features = 'sqrt',
 max_depth = 50,
 criterion = 'gini',
 bootstrap = True
)

In [122]:
rf1.fit(X_train_census, y_train_census)

RandomForestClassifier(max_depth=50, min_samples_leaf=4, min_samples_split=10,
                       n_estimators=400, random_state=5)

In [123]:
y_pred_census2 = rf1.predict(X_test_census)

In [124]:
y_pred_census1

array(['<=50K', '<=50K', '<=50K', ..., '>50K', '<=50K', '<=50K'],
      dtype=object)

In [126]:
(y_pred_census1 == y_pred_census2).sum()

np.int64(9049)

In [127]:
y_pred_census1.shape

(9049,)

In [129]:
np.array_equal(y_pred_census1, y_pred_census2)
# True since both are same, direct predict in randomizedsearchcv uses best params only

True

In [130]:
(y_pred_census == y_pred_census2).sum() #not same

np.int64(8656)

11. Answer - 85% and above

12) Which of the following Two features are most important in Random forest model?
- sensitivity and specificity
- None of the above
- Correlation between 2 trees and how strong an individual tree is
- Predict_proba

In the context of this question, the word **“features”** does **not** mean dataset columns like `age` or `education`.

Here, **“features of the Random Forest model” **means **characteristics or properties** that determine how well the model works, for example:

**Strength of individual trees** → how accurate each tree is by itself

**Correlation between trees** → how similar the trees’ predictions are

**Ability to generalize** → how well the ensemble avoids overfitting

So in this question, **“features” = important aspects of the model’s performance**, not input variables from the dataset.

The correct answer is:

**✅ Correlation between 2 trees and how strong an individual tree is**

### Explanation:

In a Random Forest:

1. **Strength of individual trees** → Each tree should perform reasonably well on its own. Strong trees improve the overall accuracy of the forest.
2. **Low correlation between trees** → Trees should make different errors. If all trees are highly correlated, the ensemble doesn’t gain much from majority voting.

Other options:

* **Sensitivity and specificity** → Metrics, not inherent features of the RF model.
* **Predict_proba** → A method to get probabilities, not a feature that affects model quality.

So the **key to Random Forest’s success** is:
**strong individual trees + low correlation among trees**.

---

12. Answer - Correlation between 2 trees and how strong an individual tree is

13) Based on what values, the feature importance will be calculated?
- Mean decrease gini and mean decrease accuracy
- All of the above
- mean increase gini and mean decrease accuracy
- mean increase gini and mean increase accuracy

The correct answer is:

**✅ Mean decrease Gini and mean decrease accuracy**

---

### Explanation:

Random Forest calculates **feature importance** based on how much each feature contributes to the model’s predictive power:

1. **Mean Decrease Gini (MDG)**

   * Measures how much a feature **decreases impurity** (Gini impurity) across all trees in the forest.
   * Features that split the data better (reduce impurity more) get higher importance.

2. **Mean Decrease Accuracy (MDA)**

   * Measures how much the **accuracy drops** when a feature’s values are randomly permuted.
   * If shuffling a feature decreases accuracy a lot → that feature is important.

---

✅ **Key point:**

* Both methods are standard in Random Forest.
* There is no concept like “mean increase Gini” or “mean increase accuracy” in feature importance calculation.

---



In [131]:
from sklearn.inspection import permutation_importance

In [132]:
# 1️⃣ Feature importance based on Gini (mean decrease impurity)
gini_importance = pd.Series(rf.feature_importances_, index=X_census.columns).sort_values(ascending=False)
print("Top features by Gini importance:")
print(gini_importance.head(10))

# 2️⃣ Feature importance based on permutation (mean decrease accuracy)
perm_importance = permutation_importance(rf, X_test_census, y_test_census, n_repeats=10, random_state=42)
perm_importance_series = pd.Series(perm_importance.importances_mean, index=X_census.columns).sort_values(ascending=False)
print("\nTop features by permutation importance:")
print(perm_importance_series.head(10))


Top features by Gini importance:
fnlwgt            0.167013
age               0.150822
capital-gain      0.119073
relationship      0.105144
education-num     0.089201
hours-per-week    0.081270
marital-status    0.068616
occupation        0.068007
workclass         0.038994
education         0.034368
dtype: float64

Top features by permutation importance:
capital-gain      0.045276
education-num     0.031064
relationship      0.026202
age               0.016720
capital-loss      0.011946
occupation        0.010764
hours-per-week    0.010090
marital-status    0.006708
workclass         0.003824
sex               0.002674
dtype: float64


Exactly 🙂 — **permutation importance is computed on test/validation data**, and here’s why:

---

### Why on the test set?

1. **Goal:** Measure how important each feature is for **generalization**, not just training.

2. **How it works:**

   * Take the trained model and the test set.
   * Shuffle (permute) the values of one feature across all test samples.
   * Compute the model’s accuracy (or other metric) on this modified test set.
   * If accuracy drops a lot → the feature is **important**.
   * If accuracy hardly changes → the feature is **less important**.

3. **Key idea:**

   * Using the **test set ensures we measure importance for unseen data**, not just the training data.
   * If we used the training set, we might overestimate importance because the model already saw those values during training.

---

* `X_test_census, y_test_census` → test set
* `n_repeats` → shuffle each feature multiple times for stability
* `perm_importance.importances_mean` → gives the mean decrease in accuracy

---

✅ **Summary:**

* **Gini importance** → training data, measures splits in trees.
* **Permutation importance** → test data, measures effect on model accuracy for unseen data.

13. Answer - Mean decrease gini and mean decrease accuracy

14) From the above model, state the disadvantage of the random forest?
- It is same as all other model
- It is a time consuming model building process
- None of the above
- It’s training time is huge due to the complexity of the model

---

## **1️⃣ Model Building**

* **Definition:** Designing and creating the structure of the machine learning model.

* **What happens:**

  * You choose the **algorithm** (e.g., Random Forest, Logistic Regression).
  * You define **hyperparameters** (e.g., `n_estimators=100`, `max_depth=10`).
  * Essentially, you **instantiate the model object** in code.

* **Example (Python, Random Forest):**

```python
from sklearn.ensemble import RandomForestClassifier

# Model building: just create the model with parameters
rf_model = RandomForestClassifier(n_estimators=100, max_depth=10, random_state=42)
```

* At this stage, **no learning has happened yet**. The model is just “empty” — ready to be trained.

---

## **2️⃣ Model Training**

* **Definition:** Actually feeding data to the model so it **learns patterns**.

* **What happens:**

  * The model sees the **training data (X_train, y_train)**.
  * It **fits** itself by finding patterns, rules, or parameters (like decision tree splits).
  * After training, the model can **predict** on new/unseen data.

* **Example (Python, Random Forest):**

```python
# Model training: fit the model on training data
rf_model.fit(X_train, y_train)
```

* Now the model has **learned from the data** and is ready for predictions.

---

### ✅ Key Difference

| Aspect          | Model Building                             | Model Training                   |
| --------------- | ------------------------------------------ | -------------------------------- |
| What it is      | Creating the model structure               | Learning from data               |
| Uses data?      | No                                         | Yes (training data)              |
| Result          | Model object with hyperparameters          | Trained model ready to predict   |
| Example in code | `RandomForestClassifier(n_estimators=100)` | `rf_model.fit(X_train, y_train)` |

---

💡 **Tip:**
Think of it like **building a car vs driving it**:

* Building = designing and assembling the car.
* Training = driving the car so it “learns the road” (metaphorically).

---


**RandomizedSearchCV** is technically **part of model building**, not training — though it involves training internally. Here’s why:

---

### 1️⃣ What RandomizedSearchCV does

* It **searches for the best hyperparameters** for your model.
* You define:

  * Which **hyperparameters to try** (`n_estimators`, `max_depth`, etc.)
  * How many **random combinations** to test (`n_iter`)
  * How to evaluate each combination (e.g., `scoring='accuracy'`)

---

### 2️⃣ How it works internally

* For each combination of hyperparameters, **RandomizedSearchCV trains the model** (fits on training data, possibly with cross-validation).
* It records the performance and picks the **best hyperparameters**.

---

### 3️⃣ Why it’s part of **model building**

* You are **still designing the “final model”** by selecting the best hyperparameters.
* Training happens **internally for evaluation**, but the **purpose is hyperparameter selection**, not directly producing the final predictive model yet.

---

### ✅ Workflow:

1. **Build model template**: `RandomForestClassifier()`
2. **Hyperparameter tuning**: `RandomizedSearchCV` → selects the best hyperparameters
3. **Train final model**: Fit the model on the full training data with the chosen hyperparameters

---

💡 **Analogy:**

* Building a cake recipe (RandomizedSearchCV tests sugar, flour, oven temp).
* Training = actually baking the cake using the chosen recipe.

---

It is a time-consuming model building process because RandomizedSearchCV only took time to fit, and it comes under model building

**But generally It’s training time is huge due to the complexity of the model**, so we go with general answer

14. Answer - It’s training time is huge due to the complexity of the model

14. ***Correct Answer*** - It is a time consuming model building process

We should have gone with our 1st answer because of RandomizeedSearchCV comes under model building


15) Which are the two methods used for hyperparameter tuning and cross-validation?
- RandomForestCLassifier
- RandomizedSearchCV and GridSearchCV
- RandomizedSearchCV
- GridSearchCV

The correct answer is:

**✅ RandomizedSearchCV and GridSearchCV**

---

### Short Description:

1. **GridSearchCV**

* Tries **all possible combinations** of specified hyperparameters.
* Performs **cross-validation** to evaluate each combination.
* Good for **small search spaces**; can be very slow if there are many hyperparameters.

2. **RandomizedSearchCV**

* Tries a **fixed number of random combinations** of hyperparameters.
* Also uses **cross-validation** for evaluation.
* Faster than GridSearchCV for **large search spaces**, but may not explore every combination.

---

💡 **Key difference:**

* **GridSearchCV** = exhaustive search
* **RandomizedSearchCV** = random sampling of hyperparameter space

Both are widely used for **hyperparameter tuning combined with cross-validation**.


15. Answer - RandomizedSearchCV and GridSearchCV

16) Use the Airport_Passengers_Satisfication.csv data for the following questions, Don’t perform any Data cleaning or pre-processing before answering the question. What percentage of Female passengers are dissatisfied?
- 44%
- 49%
- 56%
- 31%

In [133]:
airport = pd.read_csv('/content/Airport_Passengers_Satisfication.csv')

In [134]:
airport.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25976 entries, 0 to 25975
Data columns (total 24 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   id                                 25976 non-null  int64  
 1   Gender                             25976 non-null  object 
 2   Customer Type                      25976 non-null  object 
 3   Age                                25976 non-null  int64  
 4   Type of Travel                     25976 non-null  object 
 5   Class                              25976 non-null  object 
 6   Flight Distance                    25976 non-null  int64  
 7   Inflight wifi service              25976 non-null  int64  
 8   Departure/Arrival time convenient  25976 non-null  int64  
 9   Ease of Online booking             25976 non-null  int64  
 10  Gate location                      25976 non-null  int64  
 11  Food and drink                     25976 non-null  int

In [135]:
airport['Gender'].unique()

array(['Female', 'Male'], dtype=object)

In [136]:
# What percentage of Female passengers are dissatisfied?

# first filter out female passesngers

airport[airport['Gender']=='Female']

,id,Gender,Customer Type,Age,Type of Travel,Class,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,...,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes,satisfaction
0,19556,Female,Loyal Customer,52,Business travel,Eco,160,5,4,3,...,5,5,5,5,2,5,5,50,44.0,satisfied
1,90035,Female,Loyal Customer,36,Business travel,Business,2863,1,1,3,...,4,4,4,4,3,4,5,0,0.0,satisfied
4,36875,Female,Loyal Customer,49,Business travel,Eco,1182,2,3,4,...,2,2,2,2,4,2,4,0,20.0,satisfied
6,79433,Female,Loyal Customer,77,Business travel,Business,3987,5,5,5,...,5,5,5,5,4,5,3,0,0.0,satisfied
7,97286,Female,Loyal Customer,43,Business travel,Business,2556,2,2,2,...,4,4,4,4,5,4,3,77,65.0,satisfied
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
25965,15949,Female,Loyal Customer,51,Personal Travel,Eco,528,4,4,4,...,5,5,4,5,4,5,4,0,0.0,dissatisfied
25967,90347,Female,disloyal Customer,39,Business travel,Business,404,1,1,1,...,2,5,3,4,4,4,2,0,0.0,dissatisfied
25970,25309,Female,disloyal Customer,36,Business travel,Eco,432,1,5,1,...,4,5,2,5,2,3,4,0,0.0,dissatisfied
25973,37675,Female,Loyal Customer,17,Personal Travel,Eco,828,2,5,1,...,2,4,3,4,5,4,2,0,0.0,dissatisfied


In [137]:
airport[airport['Gender']=='Female']['satisfaction']

,satisfaction
0,satisfied
1,satisfied
4,satisfied
6,satisfied
7,satisfied
...,...
25965,dissatisfied
25967,dissatisfied
25970,dissatisfied
25973,dissatisfied


In [138]:
airport[airport['Gender']=='Female']['satisfaction'].value_counts(normalize=True)

,proportion
satisfaction,
dissatisfied,0.564607
satisfied,0.435393


In [142]:
airport[airport['Gender']=='Female']['satisfaction'].value_counts(
    normalize=True)['dissatisfied']

np.float64(0.5646067415730337)

In [143]:
airport[airport['Gender']=='Female']['satisfaction'].value_counts(
    normalize=True)['dissatisfied']*100

np.float64(56.46067415730337)

In [139]:
airport[airport['Gender']=='Female']['satisfaction'].value_counts(normalize=True)*100

,proportion
satisfaction,
dissatisfied,56.460674
satisfied,43.539326


In [145]:
female_dissatisfied_percentage = (
    airport[airport['Gender']=='Female']['satisfaction'].value_counts(
    normalize=True
    )*100
    )['dissatisfied']

In [146]:
female_dissatisfied_percentage

np.float64(56.46067415730337)

In [149]:
int(female_dissatisfied_percentage)

56

In [151]:
str(int(female_dissatisfied_percentage)) + '%'

'56%'

16. Answer - 56%

17) Use the Airport_Passengers_Satisfication.csv data for the following questions, Don’t perform any Data cleaning or pre-processing before answering the question. There are 14 columns that have values ranging from 1 to 5.  Take the average of the data from all such columns and assign to a new column ave_score. In the satisfaction column, convert satisfied to 1 and dissatisfied to 0. With the independent variable as ave_score and dependent variable as satisfaction, build a basic decision tree with test size=0.2 and random_state=42.  What is the accuracy_score of the model?
- 0.69
- 0.72
- 0.77
- 0.65

##### Step 1 - There are 14 columns that have values ranging from 1 to 5

In [152]:
airport.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25976 entries, 0 to 25975
Data columns (total 24 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   id                                 25976 non-null  int64  
 1   Gender                             25976 non-null  object 
 2   Customer Type                      25976 non-null  object 
 3   Age                                25976 non-null  int64  
 4   Type of Travel                     25976 non-null  object 
 5   Class                              25976 non-null  object 
 6   Flight Distance                    25976 non-null  int64  
 7   Inflight wifi service              25976 non-null  int64  
 8   Departure/Arrival time convenient  25976 non-null  int64  
 9   Ease of Online booking             25976 non-null  int64  
 10  Gate location                      25976 non-null  int64  
 11  Food and drink                     25976 non-null  int

In [156]:
airport_numeric_columns = [col for col in airport.columns if airport[col].dtype!='object']

In [158]:
len(airport_numeric_columns)

19

In [159]:
airport_numeric = airport.select_dtypes(include='number')

# Compute range for each numeric column
airport_range = airport_numeric.max() - airport_numeric.min()
print(airport_range)

id                                   129860.0
Age                                      78.0
Flight Distance                        4952.0
Inflight wifi service                     5.0
Departure/Arrival time convenient         5.0
Ease of Online booking                    5.0
Gate location                             4.0
Food and drink                            5.0
Online boarding                           5.0
Seat comfort                              4.0
Inflight entertainment                    5.0
On-board service                          5.0
Leg room service                          5.0
Baggage handling                          4.0
Checkin service                           4.0
Inflight service                          5.0
Cleanliness                               5.0
Departure Delay in Minutes             1128.0
Arrival Delay in Minutes               1115.0
dtype: float64


In [160]:
airport.describe()

,id,Age,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,Food and drink,Online boarding,Seat comfort,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes
count,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.000000,25976.00000,25893.000000
mean,65005.657992,39.620958,1193.788459,2.724746,3.046812,2.756775,2.977094,3.215353,3.261665,3.449222,3.357753,3.385664,3.350169,3.633238,3.314175,3.649253,3.286226,14.30609,14.740857
std,37611.526647,15.135685,998.683999,1.335384,1.533371,1.412951,1.282133,1.331506,1.355536,1.320090,1.338299,1.282088,1.318862,1.176525,1.269332,1.180681,1.319330,37.42316,37.517539
min,17.000000,7.000000,31.000000,0.000000,0.000000,0.000000,1.000000,0.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,1.000000,0.000000,0.000000,0.00000,0.000000
25%,32170.500000,27.000000,414.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,2.000000,3.000000,3.000000,3.000000,2.000000,0.00000,0.000000
50%,65319.500000,40.000000,849.000000,3.000000,3.000000,3.000000,3.000000,3.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,3.000000,4.000000,3.000000,0.00000,0.000000
75%,97584.250000,51.000000,1744.000000,4.000000,4.000000,4.000000,4.000000,4.000000,4.000000,5.000000,4.000000,4.000000,4.000000,5.000000,4.000000,5.000000,4.000000,12.00000,13.000000
max,129877.000000,85.000000,4983.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,5.000000,1128.00000,1115.000000


In [162]:
airport_numeric.max() # these col wil have minimum 0 also,

,0
id,129877.0
Age,85.0
Flight Distance,4983.0
Inflight wifi service,5.0
Departure/Arrival time convenient,5.0
Ease of Online booking,5.0
Gate location,5.0
Food and drink,5.0
Online boarding,5.0
Seat comfort,5.0


In [168]:
airport_numeric

,id,Age,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,Food and drink,Online boarding,Seat comfort,Inflight entertainment,On-board service,Leg room service,Baggage handling,Checkin service,Inflight service,Cleanliness,Departure Delay in Minutes,Arrival Delay in Minutes
0,19556,52,160,5,4,3,4,3,4,3,5,5,5,5,2,5,5,50,44.0
1,90035,36,2863,1,1,3,1,5,4,5,4,4,4,4,3,4,5,0,0.0
2,12360,20,192,2,0,2,4,2,2,2,2,4,1,3,2,2,2,0,0.0
3,77959,44,3377,0,0,0,2,3,4,4,1,1,1,1,3,1,4,0,6.0
4,36875,49,1182,2,3,4,3,4,1,2,2,2,2,2,4,2,4,0,20.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
25971,78463,34,526,3,3,3,1,4,3,4,4,3,2,4,4,5,4,0,0.0
25972,71167,23,646,4,4,4,4,4,4,4,4,4,5,5,5,5,4,0,0.0
25973,37675,17,828,2,5,1,5,2,1,2,2,4,3,4,5,4,2,0,0.0
25974,90086,14,1127,3,3,3,3,4,4,4,4,3,2,5,4,5,4,0,0.0


In [167]:
(airport_numeric.max() == 5)

,0
id,False
Age,False
Flight Distance,False
Inflight wifi service,True
Departure/Arrival time convenient,True
Ease of Online booking,True
Gate location,True
Food and drink,True
Online boarding,True
Seat comfort,True


In [169]:
airport_cols_with_max_5 = (airport_numeric.max() == 5)

In [172]:
airport_cols_with_max_5 = airport_cols_with_max_5[airport_cols_with_max_5]

In [174]:
airport_cols_with_max_5.index

Index(['Inflight wifi service', 'Departure/Arrival time convenient',
       'Ease of Online booking', 'Gate location', 'Food and drink',
       'Online boarding', 'Seat comfort', 'Inflight entertainment',
       'On-board service', 'Leg room service', 'Baggage handling',
       'Checkin service', 'Inflight service', 'Cleanliness'],
      dtype='object')

In [175]:
len(airport_cols_with_max_5.index)

14

In [176]:
airport_cols = airport_cols_with_max_5.index.to_list()

In [177]:
airport_cols

['Inflight wifi service',
 'Departure/Arrival time convenient',
 'Ease of Online booking',
 'Gate location',
 'Food and drink',
 'Online boarding',
 'Seat comfort',
 'Inflight entertainment',
 'On-board service',
 'Leg room service',
 'Baggage handling',
 'Checkin service',
 'Inflight service',
 'Cleanliness']

##### Step 2 - Take the average of the data from all such columns and assign to a new column ave_score.

In [179]:
airport_17 = airport.copy()

In [180]:
airport_17['ave_score'] = airport_17[airport_cols].mean(axis=1)

In [181]:
airport_17['ave_score'].describe()

,ave_score
count,25976.000000
mean,3.243439
std,0.661333
min,1.214286
25%,2.785714
50%,3.285714
75%,3.714286
max,4.928571


##### Step 3 - In the satisfaction column, convert satisfied to 1 and dissatisfied to 0.

In [183]:
airport_17['satisfaction']

,satisfaction
0,satisfied
1,satisfied
2,dissatisfied
3,satisfied
4,satisfied
...,...
25971,dissatisfied
25972,satisfied
25973,dissatisfied
25974,satisfied


In [185]:
airport_17['satisfaction'].replace('satisfied', 1, inplace=True)

In [186]:
airport_17['satisfaction'].replace('dissatisfied', 0, inplace=True)

/tmp/ipython-input-3336173086.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  airport_17['satisfaction'].replace('dissatisfied', 0, inplace=True)


In [188]:
airport_17['satisfaction']

,satisfaction
0,1
1,1
2,0
3,1
4,1
...,...
25971,0
25972,1
25973,0
25974,1


##### Step 4 - With the independent variable as ave_score and dependent variable as satisfaction, build a basic decision tree with test size=0.2 and random_state=42.

In [189]:
X_train_17, X_test_17, y_train_17, y_test_17 = train_test_split(
    airport_17[['ave_score']], airport_17['satisfaction'], test_size=0.2, random_state=42
)

In [192]:
X_train_17.shape

(20780, 1)

In [193]:
X_test_17.shape

(5196, 1)

In [194]:
y_train_17.shape

(20780,)

In [195]:
y_test_17.shape

(5196,)

In [190]:
from sklearn.tree import DecisionTreeClassifier

In [191]:
dt_17 = DecisionTreeClassifier(random_state=42)

In [197]:
dt_17.fit(X_train_17, y_train_17)

DecisionTreeClassifier(random_state=42)

##### Step 5 - What is the accuracy_score of the model?

In [198]:
y_pred_17 = dt_17.predict(X_test_17)

In [199]:
accuracy_score(y_test_17, y_pred_17)

0.722671285604311

17. Answer - 0.72

18) Use the Airport_Passengers_Satisfication.csv data for the following questions, Don’t perform any Data cleaning or pre-processing before answering the question. Build a basic random forest model with test_size=0.3, random_state=24 (parameter for both train_test_split and model initialization), n_estimators=200 with the 14 rating columns as independent variable and satisfaction column as dependent variable, what is the accuracy score of the model?
- 0.95
- 0.89
- 0.93
- 0.98

In [200]:
airport_18 = airport_17.drop(['ave_score'], axis=1)

In [201]:
X_train_18, X_test_18, y_train_18, y_test_18 = train_test_split(
    airport_18[airport_cols], airport_18['satisfaction'], test_size=0.3, random_state=24
)

In [202]:
X_train_18.shape

(18183, 14)

In [203]:
X_test_18.shape

(7793, 14)

In [204]:
y_train_18.shape

(18183,)

In [205]:
y_test_18.shape

(7793,)

In [206]:
rf_18 = RandomForestClassifier(n_estimators=200, random_state=24)

In [207]:
rf_18.fit(X_train_18, y_train_18)

RandomForestClassifier(n_estimators=200, random_state=24)

In [208]:
y_pred_18 = rf_18.predict(X_test_18)

In [209]:
accuracy_score(y_test_18, y_pred_18)

0.9358398562812781

18. Answer - 0.93

19) Use the Airport_Passengers_Satisfication.csv data for the following questions, Don’t perform any Data cleaning or pre-processing before answering the question. Build a basic random forest model in the original data set with test_size=0.3, random_state=12 (parameter for both train_test_split and model initialization) and, n_estimators=500, criterion='entropy' after performing below steps.

1. In satisfaction column, convert satisfied to 1 and dissatisfied to 0
2. Drop columns ‘id’ and 'Arrival Delay in Minutes'
3. Apply pandas get_dummies in the remaining columns.

What is the f1 score of the model?
- 0.90
- 0.98
- 0.93
- 0.95

In [231]:
airport_19 = airport.copy()

In [232]:
airport_19.columns

Index(['id', 'Gender', 'Customer Type', 'Age', 'Type of Travel', 'Class',
       'Flight Distance', 'Inflight wifi service',
       'Departure/Arrival time convenient', 'Ease of Online booking',
       'Gate location', 'Food and drink', 'Online boarding', 'Seat comfort',
       'Inflight entertainment', 'On-board service', 'Leg room service',
       'Baggage handling', 'Checkin service', 'Inflight service',
       'Cleanliness', 'Departure Delay in Minutes', 'Arrival Delay in Minutes',
       'satisfaction'],
      dtype='object')

In [233]:
len(airport_19.columns)

24

In [234]:
set(airport_19.columns) - {'id', 'Arrival Delay in Minutes', 'satisfaction'}

{'Age',
 'Baggage handling',
 'Checkin service',
 'Class',
 'Cleanliness',
 'Customer Type',
 'Departure Delay in Minutes',
 'Departure/Arrival time convenient',
 'Ease of Online booking',
 'Flight Distance',
 'Food and drink',
 'Gate location',
 'Gender',
 'Inflight entertainment',
 'Inflight service',
 'Inflight wifi service',
 'Leg room service',
 'On-board service',
 'Online boarding',
 'Seat comfort',
 'Type of Travel'}

In [235]:
len(set(airport_19.columns) - {'id', 'Arrival Delay in Minutes', 'satisfaction'})

21

In [236]:
# check number of univque values in categories

categorical_unique_count_19 = 0

for col in set(airport_19.columns) - {'id', 'Arrival Delay in Minutes', 'satisfaction'} :
  if airport_19[col].dtype == 'object' :
    categorical_unique_count_19 += airport_19[col].nunique()

categorical_unique_count_19

9

In [237]:
airport_19.drop(['id', 'Arrival Delay in Minutes', 'satisfaction'], axis=1).dtypes

,0
Gender,object
Customer Type,object
Age,int64
Type of Travel,object
Class,object
Flight Distance,int64
Inflight wifi service,int64
Departure/Arrival time convenient,int64
Ease of Online booking,int64
Gate location,int64


In [238]:
airport_19.drop(['id', 'Arrival Delay in Minutes', 'satisfaction'], axis=1).select_dtypes(
    include='object'
)

,Gender,Customer Type,Type of Travel,Class
0,Female,Loyal Customer,Business travel,Eco
1,Female,Loyal Customer,Business travel,Business
2,Male,disloyal Customer,Business travel,Eco
3,Male,Loyal Customer,Business travel,Business
4,Female,Loyal Customer,Business travel,Eco
...,...,...,...,...
25971,Male,disloyal Customer,Business travel,Business
25972,Male,Loyal Customer,Business travel,Business
25973,Female,Loyal Customer,Personal Travel,Eco
25974,Male,Loyal Customer,Business travel,Business


In [239]:
airport_19.drop(['id', 'Arrival Delay in Minutes', 'satisfaction'], axis=1).select_dtypes(
    include='object'
).nunique()

,0
Gender,2
Customer Type,2
Type of Travel,2
Class,3


In [240]:
airport_19.drop(['id', 'Arrival Delay in Minutes', 'satisfaction'], axis=1).select_dtypes(
    include='object'
).nunique().sum()

np.int64(9)

In [241]:
airport_19.columns

Index(['id', 'Gender', 'Customer Type', 'Age', 'Type of Travel', 'Class',
       'Flight Distance', 'Inflight wifi service',
       'Departure/Arrival time convenient', 'Ease of Online booking',
       'Gate location', 'Food and drink', 'Online boarding', 'Seat comfort',
       'Inflight entertainment', 'On-board service', 'Leg room service',
       'Baggage handling', 'Checkin service', 'Inflight service',
       'Cleanliness', 'Departure Delay in Minutes', 'Arrival Delay in Minutes',
       'satisfaction'],
      dtype='object')

In [242]:
len(airport_19.columns)

24

so, 24 - 2(drop id, arrival time) - 4 + 9 should be number of cols after one hot encoding. that is 27 cols

##### Step 1 - In satisfaction column, convert satisfied to 1 and dissatisfied to 0

In [244]:
airport_19['satisfaction'] = airport_19['satisfaction'].replace(
    {'satisfied' : 1, 'dissatisfied' : 0}
)

/tmp/ipython-input-2393772992.py:1: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  airport_19['satisfaction'] = airport_19['satisfaction'].replace(


In [245]:
airport_19['satisfaction']

,satisfaction
0,1
1,1
2,0
3,1
4,1
...,...
25971,0
25972,1
25973,0
25974,1


##### Step 2 - Drop columns ‘id’ and 'Arrival Delay in Minutes'

In [247]:
airport_19.drop(['id', 'Arrival Delay in Minutes'], axis=1, inplace=True)

In [248]:
airport_19.shape

(25976, 22)

##### Step 3 - Apply pandas get_dummies in the remaining columns.

In [251]:
airport_19.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25976 entries, 0 to 25975
Data columns (total 22 columns):
 #   Column                             Non-Null Count  Dtype 
---  ------                             --------------  ----- 
 0   Gender                             25976 non-null  object
 1   Customer Type                      25976 non-null  object
 2   Age                                25976 non-null  int64 
 3   Type of Travel                     25976 non-null  object
 4   Class                              25976 non-null  object
 5   Flight Distance                    25976 non-null  int64 
 6   Inflight wifi service              25976 non-null  int64 
 7   Departure/Arrival time convenient  25976 non-null  int64 
 8   Ease of Online booking             25976 non-null  int64 
 9   Gate location                      25976 non-null  int64 
 10  Food and drink                     25976 non-null  int64 
 11  Online boarding                    25976 non-null  int64 
 12  Seat

In [249]:
airport_19_encoded = pd.get_dummies(airport_19)

In [250]:
airport_19_encoded

,Age,Flight Distance,Inflight wifi service,Departure/Arrival time convenient,Ease of Online booking,Gate location,Food and drink,Online boarding,Seat comfort,Inflight entertainment,...,satisfaction,Gender_Female,Gender_Male,Customer Type_Loyal Customer,Customer Type_disloyal Customer,Type of Travel_Business travel,Type of Travel_Personal Travel,Class_Business,Class_Eco,Class_Eco Plus
0,52,160,5,4,3,4,3,4,3,5,...,1,True,False,True,False,True,False,False,True,False
1,36,2863,1,1,3,1,5,4,5,4,...,1,True,False,True,False,True,False,True,False,False
2,20,192,2,0,2,4,2,2,2,2,...,0,False,True,False,True,True,False,False,True,False
3,44,3377,0,0,0,2,3,4,4,1,...,1,False,True,True,False,True,False,True,False,False
4,49,1182,2,3,4,3,4,1,2,2,...,1,True,False,True,False,True,False,False,True,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
25971,34,526,3,3,3,1,4,3,4,4,...,0,False,True,False,True,True,False,True,False,False
25972,23,646,4,4,4,4,4,4,4,4,...,1,False,True,True,False,True,False,True,False,False
25973,17,828,2,5,1,5,2,1,2,2,...,0,True,False,True,False,False,True,False,True,False
25974,14,1127,3,3,3,3,4,4,4,4,...,1,False,True,True,False,True,False,True,False,False


In [252]:
airport_19_encoded.shape

(25976, 27)

##### Step 4 - Build a basic random forest model in the original data set with test_size=0.3, random_state=12 (parameter for both train_test_split and model initialization) and, n_estimators=500, criterion='entropy' after performing below steps.

In [253]:
X_19 = airport_19_encoded.drop(['satisfaction'], axis=1)

In [254]:
X_19.shape

(25976, 26)

In [256]:
y_19 = airport_19_encoded['satisfaction']

In [257]:
y_19

,satisfaction
0,1
1,1
2,0
3,1
4,1
...,...
25971,0
25972,1
25973,0
25974,1


In [258]:
y_19.shape

(25976,)

In [259]:
X_train_19, X_test_19, y_train_19, y_test_19 = train_test_split(
    X_19, y_19, test_size=0.3, random_state=12
)

In [260]:
X_train_19.shape

(18183, 26)

In [261]:
X_test_19.shape

(7793, 26)

In [262]:
y_train_19.shape

(18183,)

In [263]:
y_test_19.shape

(7793,)

In [264]:
rf_19 = RandomForestClassifier(n_estimators=500, criterion='entropy', random_state=12)

In [265]:
rf_19.fit(X_train_19, y_train_19)

RandomForestClassifier(criterion='entropy', n_estimators=500, random_state=12)

In [266]:
y_pred_19 = rf_19.predict(X_test_19)

In [267]:
from sklearn.metrics import f1_score

##### Step 5 - What is the f1 score of the model?

In [268]:
f1_score(y_test_19, y_pred_19)

0.9513030528667163

19. Answer - 0.95

20) Use the Airport_Passengers_Satisfication.csv data for the following questions, Don’t perform any Data cleaning or pre-processing before answering the question. Whether we can use the Random Forest method if the target variable is continuous?
- No, we can’t use continuous target variables.
- None of the options is correct
- Yes, we can use it for continuous target variables.
- Change the continuous target variable to categories and apply Random Forest

Random Forest can be used for both regression (continuous target variable) and classfication (categorical target variable)

19. ***Correct Answer*** - Change the continuous target variable to categories and apply Random Forest


The logic behind the key’s answer

Standard RandomForestClassifier requires a categorical target.

If your dataset had a continuous target, you cannot use RandomForestClassifier directly.

The “trick” answer is: convert the continuous target into categories (binning) and then apply Random Forest for classification. Because we are using Classifier version of Random Forest

20. Answer - Yes, we can use it for continuous target variables.

21) Consider we are building One Random forest with 100 trees, please choose the correct option w.r.t the process it follows?
- All columns and Random records for each tree, Build trees in parallel.
- Random columns and Random records for each tree, Build trees in series.
- Random columns and All records for each tree, Build trees in series.
- Random columns and Random records for each tree, Build trees in parallel.


The correct answer is:

**✅ Random columns and Random records for each tree, Build trees in parallel**

---

### Explanation:

1. **Random records for each tree (bagging)**

* Each tree is trained on a **random bootstrap sample** of the dataset (sampling **with replacement**).

2. **Random columns for splits**

* At each node, the tree considers a **random subset of features** (`max_features`) for splitting.
* This reduces correlation between trees and improves ensemble performance.

3. **Parallel tree building**

* Trees in a Random Forest are **independent**, so they can be built **in parallel**.
* This is why Random Forest is often faster than training many trees sequentially if parallelism is used (`n_jobs=-1` in sklearn).

---

### ✅ Key points:

* **Random records + Random columns** → reduces correlation between trees
* **Built in parallel** → independent trees allow parallel computation
* This is the essence of the **Random Forest algorithm**.


21. Answer - Random columns and Random records for each tree, Build trees in parallel.

22) Which of the following is NOT a hyperparameter in Random Forest?
- Learning rate
- Number of trees
- Maximum depth of each tree
- Number of features to consider at each split

The correct answer is:

**✅ Learning rate**

---

### Explanation:

**Random Forest hyperparameters include:**

* **Number of trees** (`n_estimators`) → how many decision trees to build
* **Maximum depth of each tree** (`max_depth`) → limits how deep each tree can grow
* **Number of features to consider at each split** (`max_features`) → controls randomness in splits

**Learning rate**

* ❌ This is **not a Random Forest hyperparameter**.
* It is a hyperparameter used in **boosting algorithms** like **Gradient Boosting** or **XGBoost**, not bagging-based Random Forest.

---

💡 **Tip:**

* Random Forest is based on **bagging**, so it does **not have a learning rate**.
* Boosting algorithms have a learning rate to **control step size** for each weak learner.


22. Answer - Learning rate